docker/Dockerfile.build

```
@@ -1,56 +1,85 @@
-# Builder image
-FROM rust:buster as builder
+# This steps implement cargo-chef for docker layer caching
+# We are using four stages:
+# - chef: installs cargo-chef
+# - planner: computes the recipe file
+# - builder: caches our dependencies and builds the binary
+# - runtime: is our runtime environment
+FROM rust:1.58.1-bullseye as chef
+RUN cargo install cargo-chef
+WORKDIR /app
+
+FROM chef AS planner
+COPY . .
+RUN cargo chef prepare  --recipe-path recipe.json
 
+FROM chef AS builder
+COPY --from=planner /app/recipe.json recipe.json
 RUN apt-get update && \
-	apt-get install -y --no-install-recommends \
-	make cmake g++ gcc llvm libclang-dev clang ca-certificates
-
-RUN mkdir /zebra
-WORKDIR /zebra
+    apt-get install -y --no-install-recommends \
+    make cmake g++ gcc llvm libclang-dev clang ca-certificates && \
+    rm -Rf /var/lib/apt/lists/* /tmp/*
+# Build dependencies - this is the caching Docker layer!
+RUN cargo chef cook --release --recipe-path recipe.json
```

If the cargo arguments don't match exactly, we will build some crates and binaries multiple times.
We should probably use a variable for the common `cargo` arguments.

```suggestion
RUN cargo chef cook --release --features enable-sentry --recipe-path recipe.json
```